(v2) Draft PR: add Pyright static typing to hook_points.py

starship006 opened the pull request:
# Description

This is a copy of PR #464 , but just remade so that @bryce13950 can close it.

-----


This draft PR adds Pyright static typing to hook_points.py. See [Alan's message on the slack](https://opensourcemechanistic.slack.com/archives/C04T79RAW8Z/p1699673111412959) for motivation/background. Specifically, this PR:
- modifies the HookPoint and HookedRootModule to support a baseline amount of pyright typing
- adds a [tool.pyright] configuration to pyproject.toml to configure the typing for this project

It's not entirely complete, but is in a state to where it would benefit from some review/assistance. Before being PR ready, we should:
- Address some TODO's I've marked on lines of code that I'm ucnertain about
- Clean up remaining type warnings on hook.py 
- If possible, adjust the pyproject.toml file so the type warnings quit applying to all files


## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] My changes generate no new warnings
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

starship006 commented:
(fixing the extra bugs right now - should be done soon)

starship006 commented:
I've changed `_HookFunctionProtocol` for the tensor parameter to take in `Union[torch.Tensor, Tuple[torch.Tensor, ...]]`, because some new code was passing in tuples of tensors into `full_hook` (i think as part of hooked SAE)? not sure if this is bad or not

added assertion statements to deal with some of these issues - though, I'm recalling that alan sent a message [here](https://github.com/TransformerLensOrg/TransformerLens/pull/464#discussion_r1456511079) about using if statements instead of assertions. is that important for our purpose (and should I go through and change all instances?)

however, I'm kind of confused on how to deal with this recurring issues. we're running into some error on calls to `isinstance(names_filter, Callable)`. i cannot figure out what is going on here, and some google searches indicate this might be a bug with the library? need some help here

bryce13950 commented on transformer_lens/hook_points.py:
```diff
@@ -523,6 +592,9 @@ def save_hook(tensor, hook, is_backward=False):
             # for attention heads the pos dimension is the third from last
+            if hook.name is None:
+                raise RuntimeError("Hook should have been provided a name")
+
```
There is a duplicate here with an assert and an if check. I prefer the if check, so if you can move that up to where the assert is. that would be great.

bryce13950 commented:
I saw that you merged main into your branch. We are actually going to be targeting dev as this will either be worked into our 2.0 release, or it will come quickly after that. I don't think there will be any new issues caused by the branch difference, but if you may want to update your branch to dev, that would probably be a good idea. 
Also, for the future, if the branch you submit is named something other than main or dev, I can actually merge the most recent origin branch into your PR without bothering you. Not a big deal by any means, but it might just save you sometime in the future.

For the asserts, from a personal preference, I do prefer having errors being checked with an if, and then thrown. Assert is for testing. It really should only be used in testing. It is pretty common to use it in the research world, but throwing errors is typically the preferred way to handle things. It allows you to be specific with your errors, and it's a lot easier to read when glancing at the code. You have to be fairly vigilant to see an assert on what is most of the time a single line as opposed to an `if` where there's a pretty deliberate message visible in an error.
That being said, this code base is filled with places where people have asserted various bits of logic. Someone needs to go through the whole code base, and replace all of these asserts with ifs at some point. That in itself will be a project. If you have the time to change the ones you have added, then I would very much appreciate. However, if we close out this PR, and we add a handful more asserts to the codebase, I don't really see that as a big deal. Maybe we are adding 2-3% more work to the eventual project of changing them everywhere. 
The way I see it is that you have been waiting to get this change into the code base for over 5 months. If you have the time to change them, then that's great, but I am not going to require that you spend the time to do that after all of this time when you are basically following the existing style.

bryce13950 commented:
For Hooked SAE stuff, don't worry about that. We are removing that in 2.0 in favor of sending people over to SAELens. If you want to wait until tomorrow, I can have it removed, so you can merge agains the pending 2.0 with it about 98% ready for release.

starship006 commented:
Okay sounds good, I can wait until the TLens update and then I'll make these changes

bryce13950 commented:
Everything pending for 2.0 is in the `dev` branch. If you want to just update to that branch, you should be ready to wrap this up. We are at the moment just waiting for something to be added to another project, but the actual 2.0 at this moment is going to change very little.

bryce13950 commented:
I'll have the hooked SAE stuff removed tomorrow, but you can just ignore it if it's still there.

starship006 commented:
I reverted the change to `_HookFunctionProtocol` and moved up the if statement.

I think I'm being a git noob herem but I can't figure out how to merge the dev branch into this, all I can do is merge main into it

starship006 commented:
> Someone needs to go through the whole code base, and replace all of these asserts with ifs at some point. That in itself will be a project. If you have the time to change the ones you have added, then I would very much appreciate it. However, if we close out this PR, and we add a handful more asserts to the codebase, I don't really see that as a big deal. Maybe we are adding 2-3% more work to the eventual project of changing them everywhere.

I left the other ones alone for now, I want to get this PR done first. I might be able to do a second pass-through and clean them all up down the line.

bryce13950 commented:
If you can rename your pr branch, I can get dev merged into your branch, and wrap up the last minute details on this.

starship006 commented:
did that, thanks
